pkg/gateway/multiparts/tracker.go

```diff
@@ -38,82 +46,70 @@ var (
 	ErrInvalidMetadataSrcFormat = errors.New("invalid metadata source format")
 )
 
-func NewTracker(adb db.Database) Tracker {
+func NewTracker(ms kv.StoreMessage) Tracker {
 	return &tracker{
-		db: adb,
+		store: ms,
+	}
+}
+
+func multipartFromProto(pb *MultipartUploadData) *MultipartUpload {
+	return &MultipartUpload{
+		UploadID:        pb.UploadId,
+		Path:            pb.Path,
+		CreationDate:    pb.CreationDate.AsTime(),
+		PhysicalAddress: pb.PhysicalAddress,
+		Metadata:        pb.Metadata,
+		ContentType:     pb.ContentType,
+	}
+}
+
+func protoFromMultipart(m *MultipartUpload) *MultipartUploadData {
+	return &MultipartUploadData{
+		UploadId:        m.UploadID,
+		Path:            m.Path,
+		CreationDate:    timestamppb.New(m.CreationDate),
+		PhysicalAddress: m.PhysicalAddress,
+		Metadata:        m.Metadata,
+		ContentType:     m.ContentType,
 	}
 }
 
 func (m *tracker) Create(ctx context.Context, multipart MultipartUpload) error {
 	if multipart.UploadID == "" {
 		return ErrInvalidUploadID
 	}
-	_, err := m.db.Transact(ctx, func(tx db.Tx) (interface{}, error) {
-		_, err := tx.Exec(`INSERT INTO gateway_multiparts (upload_id,path,creation_date,physical_address, metadata, content_type)
-			VALUES ($1, $2, $3, $4, $5, $6)`,
-			multipart.UploadID, multipart.Path, multipart.CreationDate, multipart.PhysicalAddress, multipart.Metadata, multipart.ContentType)
-		return nil, err
-	})
+	_, err := m.Get(ctx, multipart.UploadID)
+	if err == nil {
+		return ErrAlreadyExists
+	}
+	path := kv.NewPath(multipartPrefix, multipart.UploadID)
+	err = m.store.SetMsg(ctx, path, protoFromMultipart(&multipart))
 	return err
 }
 
 func (m *tracker) Get(ctx context.Context, uploadID string) (*MultipartUpload, error) {
 	if uploadID == "" {
 		return nil, ErrInvalidUploadID
 	}
-	res, err := m.db.Transact(ctx, func(tx db.Tx) (interface{}, error) {
-		var m MultipartUpload
-		if err := tx.Get(&m, `
-			SELECT upload_id, path, creation_date, physical_address, metadata, content_type 
-			FROM gateway_multiparts
-			WHERE upload_id = $1`,
-			uploadID); err != nil {
-			return nil, err
-		}
-		return &m, nil
-	})
+	data := &MultipartUploadData{}
+	path := kv.NewPath(multipartPrefix, uploadID)
+	err := m.store.GetMsg(ctx, path, data)
 	if err != nil {
 		return nil, err
 	}
-	return res.(*MultipartUpload), nil
+	return multipartFromProto(data), nil
 }
 
 func (m *tracker) Delete(ctx context.Context, uploadID string) error {
 	if uploadID == "" {
 		return ErrInvalidUploadID
 	}
-	_, err := m.db.Transact(ctx, func(tx db.Tx) (interface{}, error) {
-		res, err := tx.Exec(`DELETE FROM gateway_multiparts WHERE upload_id = $1`, uploadID)
-		if err != nil {
-			return nil, err
+	if _, err := m.Get(ctx, uploadID); err != nil {
```

Comment:
I would not implement Delete in the tracker level, using Get in the tracker level - until there is another lower implementation of 'get'.
This is because usually the logging, BI, etc will be implemented at that level.
In this above case - I'll use the storage Get and not the upper level one.